datastore/gcloud/aio/datastore/filter.py

```diff
@@ -0,0 +1,111 @@
+from typing import Any
+from typing import Dict
+from typing import List
+
+from gcloud.aio.datastore.constants import CompositeFilterOperator
+from gcloud.aio.datastore.constants import PropertyFilterOperator
+from gcloud.aio.datastore.value import Value
+
+
+class _BaseFilter:
```

Comment:
Huh, I've never seen abstract base classes with a leading underscore before. We've just used `BaseClassName` everywhere else, so consistency would be great here, but if there's a good reason for the leading underscore, we'd be happy to use this model?

Reply:
Ah sorry, didn't see that there was a convention.

It is partly because the class is abstract but more because it's internal to the library. I just wanted it to look wrong if somebody tries to do `filter = _BaseFilter()`. A bit like trying to access `foo._bar()`

But now that you point it out, this isn't actually necessary since it's not exposed to the outside world. I still like the fact that `filter = _BaseFilter()` looks wrong. Anyway, I think it's better to follow the convention so I'll make the change.

As a side note, I read a bit about abstract methods in Python and found out that there is something called [abc](https://docs.python.org/3.7/library/abc.html) that could be useful if you wanted to be strict about that kind of thing. (See also [PEP3119](https://www.python.org/dev/peps/pep-3119/))

Reply:
`abc` is neat, but the thing I think would be most useful about marking off `ABC`s isn't handled by it (ie. preventing instantiation of the base class) :( I suppose it might be worth adding an `__init__()` which raises `NotImplemented` to handle that...

```
>>> import abc
>>> class Foo(abc.ABC):
...   pass
...
>>> x = Foo()
>>> x
<__main__.Foo object at 0x108bc4b38>
```

Reply:
Ah, smart. I prefer to do that in __new__ rather than __init__, but that's a good way to make it explicit. Will make the change.

Reply:
New could make sense for sure, if that seems cleaner!